nibabel/pointset.py

```diff
@@ -0,0 +1,256 @@
+"""Point-set structures
+
+Imaging data are sampled at points in space, and these points
+can be described by coordinates.
+These structures are designed to enable operations on sets of
+points, as opposed to the data sampled at those points.
+
+Abstractly, a point set is any collection of points, but there are
+two types that warrant special consideration in the neuroimaging
+context: grids and meshes.
+
+A *grid* is a collection of regularly-spaced points. The canonical
+examples of grids are the indices of voxels and their affine
+projection into a reference space.
+
+A *mesh* is a collection of points and some structure that enables
+adjacent points to be identified. A *triangular mesh* in particular
+uses triplets of adjacent vertices to describe faces.
+"""
+from __future__ import annotations
+
+import math
+import typing as ty
+from dataclasses import dataclass, replace
+
+import numpy as np
+
+from nibabel.casting import able_int_type
+from nibabel.fileslice import strided_scalar
+from nibabel.spatialimages import SpatialImage
+
+if ty.TYPE_CHECKING:  # pragma: no cover
+    from typing_extensions import Self
+
+    _DType = ty.TypeVar('_DType', bound=np.dtype[ty.Any])
+
+
+class CoordinateArray(ty.Protocol):
+    ndim: int
+    shape: tuple[int, int]
+
+    @ty.overload
+    def __array__(self, dtype: None = ..., /) -> np.ndarray[ty.Any, np.dtype[ty.Any]]:
+        ...  # pragma: no cover
+
+    @ty.overload
+    def __array__(self, dtype: _DType, /) -> np.ndarray[ty.Any, _DType]:
+        ...  # pragma: no cover
+
+
+@dataclass
+class Pointset:
+    """A collection of points described by coordinates.
+
+    Parameters
+    ----------
+    coords : array-like
+      2-dimensional array with coordinates as rows
+    affine : :class:`numpy.ndarray`
+      Affine transform to be applied to coordinates array
+    homogeneous : :class:`bool`
+      Indicate whether the provided coordinates are homogeneous,
+      i.e., homogeneous 3D coordinates have the form ``(x, y, z, 1)``
+    """
+
+    coordinates: CoordinateArray
+    affine: np.ndarray
+    homogeneous: bool = False
+    ndim = 2
+    __array_priority__ = 99
+
+    def __init__(
+        self,
+        coordinates: CoordinateArray,
+        affine: np.ndarray | None = None,
+        homogeneous: bool = False,
+    ):
+        self.coordinates = coordinates
+        self.homogeneous = homogeneous
+
+        if affine is None:
+            self.affine = np.eye(self.dim + 1)
+        else:
+            self.affine = np.asanyarray(affine)
+
+        if self.affine.shape != (self.dim + 1,) * 2:
+            raise ValueError(f'Invalid affine for {self.dim}D coordinates:\n{self.affine}')
+        if np.any(self.affine[-1, :-1] != 0) or self.affine[-1, -1] != 1:
+            raise ValueError(f'Invalid affine matrix:\n{self.affine}')
+
+    @property
+    def shape(self) -> tuple[int, int]:
+        """The shape of the coordinate array"""
+        return self.coordinates.shape
```

Comment:
This is the same as `n_coords`. I would say we should have one or the other, but not both.

Reply:
And note that `n_coords` is comparable to `n_triangles` when this is subclassed into `TriangularMesh`, so I'm a little partial to `n_coords`, but if you have a thought that would work for both coordinates and triangles, happy to consider it.

Reply:
My suggestion only operates under the logic of making the PointSet more numpy-like. I guess a TriangularMesh should not define the size property. But the same question could be posed for `shape` then. Why does `shape` conceptually refer to the vertices but size does not?

As a note, if some time we support other meshes with more than three edges, then `n_triangles` should be better named `n_faces`.

(btw, not particularly attached to this suggestion, so happy to just let go)

Reply:
I think I was taking the array-like interface a little too seriously at one point. I agree shape isn't really needed, so I'm happy to drop both `shape` and `size`. It can be added back later with better motivation, if there is one.